sim/paladin/som.go

```diff
@@ -0,0 +1,101 @@
+		ActionID:    core.ActionID{SpellID: 407803}, // Judgement of Righteousness.
+		SpellSchool: core.SpellSchoolHoly,
+		ProcMask:    core.ProcMaskMeleeSpecial,
+		Flags:       core.SpellFlagMeleeMetrics | SpellFlagSecondaryJudgement,
+
+		DamageMultiplier: 0.7,
+		CritMultiplier:   paladin.MeleeCritMultiplier(),
+		ThreatMultiplier: 1,
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			baseDamage := spell.Unit.MHWeaponDamage(sim, spell.MeleeAttackPower()) + spell.BonusWeaponDamage()
+			spell.CalcAndDealDamage(sim, target, baseDamage, spell.OutcomeMeleeSpecialCritOnly)
+		},
+	})
+
+	onSwingProc := paladin.RegisterSpell(core.SpellConfig{
+		ActionID:         core.ActionID{SpellID: 407799},
+		SpellSchool:      core.SpellSchoolHoly,
+		ProcMask:         core.ProcMaskMeleeMHSpecial | core.ProcMaskSuppressedExtraAttackAura,
+		Flags:            core.SpellFlagMeleeMetrics,
+		RequiredLevel:    1,
+		DamageMultiplier: 0.4,
+		ThreatMultiplier: 1.0,
+		CritMultiplier:   paladin.MeleeCritMultiplier(),
+
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			baseDamage := spell.Unit.MHNormalizedWeaponDamage(sim, spell.MeleeAttackPower()) + spell.BonusWeaponDamage()
+			spell.CalcAndDealDamage(sim, target, baseDamage, spell.OutcomeMeleeSpecialHitAndCrit)
+		},
+	})
+
+	auraActionID := core.ActionID{SpellID: 407798}
+	paladin.SealOfMartyrdomAura = paladin.RegisterAura(core.Aura{
+		Label:    "Seal of Martyrdom",
+		Tag:      "Seal",
+		ActionID: auraActionID,
+		Duration: SealDuration,
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if !result.Landed() {
+				return
+			}
+
+			if spell.ProcMask.Matches(core.ProcMaskMeleeWhiteHit) {
+				onSwingProc.Cast(sim, result.Target)
+			}
+			if spell.ProcMask.Matches(core.ProcMaskProc) {
+				onSwingProc.Cast(sim, result.Target)
+			}
+			if spell.Flags.Matches(SpellFlagPrimaryJudgement) {
+				onJudgementProc.Cast(sim, result.Target)
+			}
+		},
+	})
+	// Necessary because of the mix of % base mana cost and flat reduction on the libram
+	manaCost := paladin.BaseMana * 0.04
+	if paladin.Ranged().ID == LibramOfBenediction {
+		manaCost -= 10
+	}
+	aura := paladin.SealOfMartyrdomAura
+	paladin.SealOfMartyrdom = paladin.RegisterSpell(core.SpellConfig{
+		ActionID:    auraActionID,
+		SpellSchool: core.SpellSchoolHoly,
+		Flags:       core.SpellFlagAPL,
+
+		ManaCost: core.ManaCostOptions{
+			FlatCost:   manaCost,
+			Multiplier: 1.0 - (float64(paladin.Talents.Benediction) * 0.03),
+		},
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				GCD: core.GCDDefault,
+			},
+		},
+
+		ApplyEffects: func(sim *core.Simulation, _ *core.Unit, spell *core.Spell) {
+			if paladin.CurrentSeal != nil {
+				paladin.CurrentSeal.Deactivate(sim)
+			}
+			paladin.CurrentSeal = aura
+			paladin.CurrentSeal.Activate(sim)
+			paladin.CurrentSealExpiration = sim.CurrentTime + SealDuration
```

Comment:
This looks like something you do in all seal spells. Can you make it a generic function in Paladin something like:
```
func (paladin *Paladin) ApplySeal(aura *core.Aura) {
	if paladin.CurrentSeal != nil {
		paladin.CurrentSeal.Deactivate(sim)
	}
	paladin.CurrentSeal = aura
	paladin.CurrentSeal.Activate(sim)
	paladin.CurrentSealExpiration = sim.CurrentTime + aura.Duration
}
```
And then call this from all seal ApplyEffects

Reply:
Implemented in https://github.com/wowsims/sod/pull/302/commits/0663be7b6d346fbe5b525568cde39b19ddbb44db